src/clr-angular/data/datagrid/providers/column-reorder.service.ts

```diff
@@ -0,0 +1,122 @@
+      // Dragged to the right so each in-between columns should decrement their flex orders
+      for (let i = draggedFrom + 1; i <= draggedTo; i++) {
+        this.reorderQueue[i] = i - 1;
+      }
+    } else if (draggedTo < draggedFrom) {
+      // Dragged to the left so each in-between columns should decrement their flex orders
+      for (let i = draggedFrom - 1; i >= draggedTo; i--) {
+        this.reorderQueue[i] = i + 1;
+      }
+    }
+    this.reorderQueue[draggedFrom] = draggedTo;
+    // After queueing all required specs, emit what kind of reorder is requested
+    if (this.reorderQueue.length > 0) {
+      this._reorderRequested.next(this.reorderQueue);
+    }
+  }
+
+  private hasDiffWith(newOrders: number[]): boolean {
+    if (!this.orders) {
+      return true;
+    }
+    if (!newOrders) {
+      return false;
+    }
+    if (this.orders.length !== newOrders.length) {
+      return true;
+    }
+    return newOrders.filter((newOrder, index) => newOrder !== this.orders[index]).length > 0;
+  }
+
+  // ClrDatagrid updates this array of order values after re-ordering happens.
+  // Then, rows will change their cell position according to its order values in this array.
+  // So when re-ordering happens, we broadcast through "reorderCompleted" subject to notify rows.
+  orders: number[];
+
+  containerRef: ViewContainerRef;
+
+  get columnsGroupId() {
+    return this._columnsGroupId;
+  }
+
+  get reorderRequested(): Observable<number[]> {
+    return this._reorderRequested.asObservable();
+  }
+
+  get reorderCompleted(): Observable<void> {
+    return this._reorderCompleted.asObservable();
+  }
+
+  // The following method is called by column ClrDatagridColumn when one column is dropped on another.
+  public reorderViews(draggedView: ViewRef, targetView: ViewRef): void {
+    const draggedFrom = this.containerRef.indexOf(draggedView);
+    const draggedTo = this.containerRef.indexOf(targetView);
+    this.processReorderRequest(draggedFrom, draggedTo);
+  }
+
+  // The following method will be called by ClrDatagrid, after it finishes applying order changes, to broadcast to
+  // all cells how column orders are updated and then cells will start applying changed orders.
+  public updateOrders(orders: number[], onReordering = false): void {
+    if (orders && orders.length > 0 && this.hasDiffWith(orders)) {
+      // update with new orders
+      this.orders = orders;
+
+      // update the columns states with their new orders
+      this.orders.forEach((order, index) => {
+        this.columnsService.emitStateChangeAt(index, { changes: [DatagridColumnChanges.ORDER], order: order });
+      });
+
+      // notify whatever that needs to respond to reorder changes
+      if (onReordering) {
+        this._reorderCompleted.next();
+      }
+
+      this.columnsService.requestFirstVisibleChangeCheck();
+      this.columnsService.requestLastVisibleChangeCheck();
+    }
+  }
+
+  public orderAt(index: number): number {
```

Comment:
```js
function orderAt(index: number): number {
  return (this.orders || [])[index] || -1
}
```
Talking to the Duck: If `this.order` is not existing use empty array, try to find an item in this array by index, If I find one, return it to me else return `-1`

* `this.orders` must always be an array of numbers - no need to check it - if there is some way to put something that is not a number this must be handled at that place.
* `this.orders` is initialized at the start with empty array we could remove this `(this.orders || [])`

it seems that this a use case here .. so maybe an initial value may be skipped
```js
private hasDiffWith(newOrders: number[]): boolean {
    if (!this.orders) {
      return true;
    }
    ...
```

Reply:
```ts
function orderAt(index: number): number {
  return (this.orders || [])[index] || -1
}
```
this approach would return -1 if the `this.orders[index]` returns 0.

Reply:
My Duck did not expect to have `0` in the list :) So I did not test for that case. But you are right.

```js
var orders = [0,2,0,4]
function foo(index) { return (orders || [])[index] || -1}
foo(2)
// => -1
```